Update azure_core README (#2174)
Azure/azure-sdk-for-rust: sdk/core/azure_core/README.md

# Azure Core shared client library for Rust

`azure_core` provides shared primitives, abstractions, and helpers for modern Rust Azure SDK client libraries.
These libraries follow the [Azure SDK Design Guidelines for Rust](https://azure.github.io/azure-sdk/rust_introduction.html)

> and can be easily identified by package and namespaces names starting with 'azure_', e.g. `azure_identity`.

Reviewer: ```suggestion
and can be easily identified by package and namespaces names starting with `azure_`, e.g. `azure_identity`.
```

`azure_core` allows client libraries to expose common functionality in a consistent fashion,
so that once you learn how to use these APIs in one client library, you will know how to use them in other client libraries.

[Source code] | [Package (crates.io)] | [API Reference Documentation]

## Getting started

Typically, you will not need to install `azure_core`;
it will be installed for you when you install one of the client libraries using it.
In case you want to install it explicitly (to implement your own client library, for example),
you can find the crates.io package [here][Package (crates.io)].

## Key concepts

The main shared concepts of `azure_core` (and so Azure SDK libraries using `azure_core`) include:

> - Configuring service clients, e.g. configuring retries, logging (`ClientOptions`).
> - Accessing HTTP response details (`Response`, `Response<T>`).
> - Calling long-running operations (`Operation<T>`).
> - Paging and asynchronous streams (`AsyncPageable<T>`).
> - Exceptions for reporting errors from service requests in a consistent fashion. (`RequestFailedException`).
> - Customizing requests (`RequestContext`).
> - Abstractions for representing Azure SDK credentials. (`TokenCredentials`).

Reviewer: Most of this aren't our concepts, but .NET's it looks like.
```suggestion
- Configuring service clients, e.g. configuring retries, logging (`ClientOptions`).
- Accessing HTTP response details (`Response<T>`).
- Calling long-running operations (`Poller<T>`).
- Paging and asynchronous streams (`Pager<T>`).
- Errors from service requests in a consistent fashion. (`azure_core::Error`).
- Customizing requests (`ClientMethodOptions`).
- Abstractions for representing Azure SDK credentials. (`TokenCredentials`).
```

Reviewer: Headers/topics below would need to match.

[Long-running operations](#consuming-long-running-operations-using-operationt) |
<!-- [Handling failures](#reporting-errors-requestfailedexception) |
[Diagnostics](https://github.com/Azure/azure-sdk-for-rust/blob/main/sdk/core/azure_core/samples/Diagnostics.md) |
[Mocking](https://learn.microsoft.com/rust/azure/sdk/unit-testing-mocking) |
[Client lifetime](https://devblogs.microsoft.com/azure-sdk/lifetime-management-and-thread-safety-guarantees-of-azure-sdk-rust-clients/) -->
<!-- CLIENT COMMON BAR -->

## Examples

**NOTE:** Samples in this file apply only to packages that follow [Azure SDK Design Guidelines](https://azure.github.io/azure-sdk/rust_introduction.html). Names of such packages usually start with `azure_`.

### Configuring Service Clients Using `ClientOptions`

Azure SDK client libraries typically expose one or more _service client_ types that
are the main starting points for calling corresponding Azure services.
You can easily find these client types as their names end with the word _Client_.
For example, `SecretClient` can be used to call the Key Vault service and interact with secrets,
and `KeyClient` can be used to access Key Vault service cryptographic keys.

> These client types can be instantiated by calling a simple `new` or `builder` function that takes various configuration options.These options are passed as a parameter that extends `ClientOptions` class exposed by `azure_core`.

Reviewer: I wouldn't mention builders. Only a few will use it, but it wouldn't be common enough to mention in `azure_core` I think.

Author: Removed mention builders

Various service specific options are usually added to its subclasses, but a set of SDK-wide options are
available directly on `ClientOptions`.

```rust no_run
use azure_core::ClientOptions;
use azure_identity::DefaultAzureCredential;
use azure_security_keyvault_secrets::{SecretClient, SecretClientOptions};

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let options = SecretClientOptions {

>         api_version: "7.0".to_string(),

Reviewer: That's *really* told. Perhaps something more recent, like the latest stable:
```suggestion
        api_version: "7.5".to_string(),
```

_Service clients_ have methods that can be used to call Azure services. We refer to these client methods as _service methods_.
_Service methods_ return a shared `azure_core` type `Response<T>` (in rare cases its non-generic sibling, a raw `Response`).
This type provides access to both the deserialized result of the service call, and to the details of the HTTP response returned from the server.

```rust no_run
use azure_core::Response;
use azure_identity::DefaultAzureCredential;
use azure_security_keyvault_secrets::{models::SecretBundle, SecretClient};

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    // create a client
    let credential = DefaultAzureCredential::new()?;
    let client = SecretClient::new(
        "https://your-key-vault-name.vault.azure.net/",
        credential.clone(),
        None,
    )?;

    // call a service method, which returns Response<T>
    let response: Response<SecretBundle> = client.get_secret("SecretName", "", None).await?;

    // Response<T> has two main accessors.
    // The into_body() function for accessing the deserialized result of the call
    let secret = response.into_body().await?;

>         // get response again because it was moved in above statement
>     let response: Response<SecretBundle> = client.get_secret("SecretName", "", None).await?;

Reviewer: ```suggestion
    // get response again because it was moved in above statement
    let response: Response<SecretBundle> = client.get_secret("SecretName", "", None).await?;
```

    // .. and the deconstruct() method for accessing all the details of the HTTP response

>     let (status, headers, body) = response.deconstruct();

Reviewer: This won't work. `into_body()` takes ownership of `response` i.e., it takes a `self` instead of `&self`. So does `deconstruct`. It's one or the other. Most people won't need/use `deconstruct`.

Author: I think this section in the .NET README wanted to show how to get the HTTP response details. Is there a better way than using deconstruct.

Author: @heaths, won't redefining the response by fetching it again fix that?

    // for example, you can access HTTP status
    println!("Status: {}", status);

    // or the headers
    for (header_name, header_value) in headers.iter() {
        println!("{}: {}", header_name.as_str(), header_value.as_str());
    }

    Ok(())
}
```

More on response types in [response samples](https://github.com/Azure/azure-sdk-for-rust/blob/main/sdk/core/azure_core/samples/Response.md).

### Setting up console logging

To create an Azure SDK log listener that outputs messages to console use `AzureEventSourceListener::create_console_logger` method.

> ```rust no_run
> // Setup a listener to monitor logged events.
> let listener = AzureEventSourceListener::create_console_logger();
> ```

Reviewer: That won't work. That's .NET's listenders. We use the `tracing` crate, so you can use a `tracing-subscriber` e.g.,
```suggestion
```rust no_run
// Setup a subscriber to listen for trace events.
tracing_subscriber::fmt::init();
```
```

> More on logging in [diagnostics samples](https://github.com/Azure/azure-sdk-for-rust/blob/main/sdk/core/azure_core/samples/Diagnostics.md).

Reviewer: We don't have such a page.

Author: Removed from README

> ### Reporting Errors `RequestFailedException`

Reviewer: As noted above, no exceptions.

Author: Altered section to be on Handling Result of Error

Mocking is enabled by:

- providing a protected parameterless constructor on client types.
- making service methods virtual.
- providing APIs for constructing model types returned from virtual service methods. To find these factory methods look for types with the _ModelFactory_ suffix, e.g. `SecretModelFactory`.

For example, the ConfigurationClient.get method can be mocked (with [Mockall](https://github.com/asomers/mockall)) as follows:

```rust no_run
// Create a mock response
let mock_response = MockResponse::new();

// Create a mock value
let mock_value = SecretModelFactory::key_vault_secret(
    SecretModelFactory::secret_properties("http://example.com".parse().unwrap())
);

// Create a client mock
let mut mock = MockSecretClient::new();

// Setup client method
mock.expect_get_secret()
    .with(eq("Name"), eq(None), eq(Default::default()))
    .returning(move |_, _, _| Ok(Response::from_value(mock_value.clone(), mock_response.clone())));

// Use the client mock
let client = mock;
let secret = client.get_secret("Name").await?;
```

More on mocking in [Unit testing and mocking with the Azure SDK for Rust](https://learn.microsoft.com/rust/azure/sdk/unit-testing-mocking).

> ## Distributed tracing with OpenTelemetry

Reviewer: This whole section is incorrect and we should probably just elide for now. We do support `tracing` which has an `opentelemetry` subscriber, but we haven't tested yet and don't actually have *distributed* tracing wired up correctly yet. I already have tracking issues open.

Author: Removed the section on tracing for now.
